### DISCUSSION TITLE
How to create a label progromaticaly?

### DISCUSSION THREAD

**Discussion by ainsleyclark** [terrastruct/d2#836] [General]
Hi there,
I'm just wondering how we can create a label programmatically using the library?
The below doesn't work, I think I'm missing something.

```go
_, _, err = d2oracle.Create(graph, "test1 -> test2:hey")
if err != nil {
	return err
}
```

**Comment by alixander** [comment]
hey @ainsleyclark , I'll make a note to document this better, but the best place to look right now is our tests, which we have a large number of.

For example, edge labels are found here:

https://github.com/terrastruct/d2/blob/d369371771ee76c75d052ec13a11a9e7771d59b4/d2oracle/edit_test.go#L1010

And you can scroll down to how this is tested:

https://github.com/terrastruct/d2/blob/d369371771ee76c75d052ec13a11a9e7771d59b4/d2oracle/edit_test.go#L1164

So the correct usage is:

1. Create the edge: `d2oracle.Create(graph, "test1 -> test2")`
2. Set the label by targeting the edge: `d2oracle.Set(graph, "(test1 -> test2)[0]", "hey")`

**Comment by ainsleyclark** [comment]
Hey @alixander 

Thanks for your swift response.

I think I maybe missing something:

```go
_, graph, _ := d2lib.Compile(context.Background(), "", nil)

_, _, err := d2oracle.Create(graph, "test1 -> test2")
if err != nil {
	log.Fatal(err)
}

t2 := "hey"
_, err = d2oracle.Set(graph, "(test1 -> test2)[0]", &t2, nil)
// Also tried _, err = d2oracle.Set(graph, "(test1 -> test2)[0]", nil, &t2)
if err != nil {
	log.Fatal(err)
}
```

Prints:
```
 Diagram.Render  "2023/02/17 14:40:47 failed to set \"(test1 -> test2)[0]\" to \"hey\" \"(*string)(nil)\": edge not found"
```

Just out of interest, are there any docs on using the library in `go`? 
I've read [this](https://terrastruct.com/blog/post/generate-diagrams-programmatically/) tutorial which is great but I would love to see a deep dive in how to use this a little more with styles etc.

Thanks again.

**Reply by alixander** [reply]
@ainsleyclark Oh, if you look at the signature for those methods, they return a graph. So the functions don't mutate the original graph passed in.

There's no docs right now, on the near-term queue to write some. But that doesn't mean it's immature -- the library is used extensively every day  on Terrastruct (we use it to change D2 diagrams when users do things on the GUI, like reconnect an arrow). Out of curiosity, what's your use case?

**Comment by ainsleyclark** [comment]
Ah perfect thanks for answering that, it's working perfectly now.

Some docs on the `d2oracle` package would be amazing, even if it's go doc referencing the public methods like `Rename`, `Set` etc as they're not immediately obvious what they do / when you should use them

I'm constructing a fairly big infrastructure diagram of events produced/consumed within AWS. Which kind of leads me to my next question (appreciate your patience), is there a way to make all connecting lines straight? Otherwise, I may get in a tangled mess when I run this over around 100 microservices; any advice for this sort of large scale diagram would be great.

Thanks once again.

**Reply by alixander** [reply]
V1 of some docs: https://d2lang.com/tour/api

For over 100 interconnected microservices, it's likely that a force-directed layout is needed. They handle large diagrams way better than hierarchical layouts, and use straight lines. D2 unfortunately has yet to integrate such an engine -- it's still in the TODO phase: https://github.com/terrastruct/d2/issues/641 .

If there is some way to encapsulate chunks of those 100, you will benefit from multi-board diagrams that lets users click **into** parts of the diagram (https://github.com/terrastruct/d2/issues/759). This splits a huge diagram that's hard to comprehend, into multiple smaller ones that are easier to consume. This is also a wip right now, but 95% of the work here is done. If this is feasible, it's a short wait -- but if not, you might need to revisit in a month or so whenever that force-directed layout is done. (any interest in writing a layout engine? 😄 )

**Reply by ainsleyclark** [reply]
Ah I missed those docs! Thanks for pointing me in the right direction.

Ah that's a shame, but I can imagine it's a lot of wok. I could foresee that wiring 100+ objects together might be a bit spaghetti like without the force-directed layout. I think splitting them up (multi-board) may be off the cards for us as we need to see the full picture & zoom in if necessary.

I would love to help out! But I'm afraid I don't have the time at the minute :(
